Content.Server/Shuttles/DockingSystem.cs

```diff
@@ -0,0 +1,463 @@
+using System;
+using Content.Server.Doors.Components;
+using Content.Server.Power.Components;
+using Content.Shared.Doors;
+using Content.Shared.Shuttles;
+using Content.Shared.Verbs;
+using Robust.Shared.GameObjects;
+using Robust.Shared.IoC;
+using Robust.Shared.Localization;
+using Robust.Shared.Log;
+using Robust.Shared.Map;
+using Robust.Shared.Maths;
+using Robust.Shared.Physics;
+using Robust.Shared.Physics.Collision.Shapes;
+using Robust.Shared.Physics.Dynamics;
+using Robust.Shared.Physics.Dynamics.Joints;
+using Robust.Shared.Utility;
+using Robust.Shared.ViewVariables;
+
+namespace Content.Server.Shuttles
+{
+    [RegisterComponent]
+    public sealed class DockingComponent : SharedDockingComponent
+    {
+        [ViewVariables]
+        public DockingComponent? DockedWith;
+
+        [ViewVariables]
+        public Joint? DockJoint;
+
+        [ViewVariables]
+        public override bool Docked => DockedWith != null;
+    }
+
+    public sealed class DockingSystem : EntitySystem
+    {
+        [Dependency] private readonly IMapManager _mapManager = default!;
+        [Dependency] private readonly SharedBroadphaseSystem _broadphaseSystem = default!;
+        [Dependency] private readonly SharedJointSystem _jointSystem = default!;
+
+        private const string DockingFixture = "docking";
+        private const string DockingJoint = "docking";
+        private const float DockingRadius = 0.20f;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            SubscribeLocalEvent<DockingComponent, ComponentStartup>(OnStartup);
+            SubscribeLocalEvent<DockingComponent, ComponentShutdown>(OnShutdown);
+            SubscribeLocalEvent<DockingComponent, PowerChangedEvent>(OnPowerChange);
+            SubscribeLocalEvent<DockingComponent, AnchorStateChangedEvent>(OnAnchorChange);
+
+            SubscribeLocalEvent<DockingComponent, GetInteractionVerbsEvent>(OnVerb);
+            SubscribeLocalEvent<DockingComponent, BeforeDoorAutoCloseEvent>(OnAutoClose);
+            SubscribeLocalEvent<DockingComponent, DoorOpenAttemptEvent>(OnDoorOpenAttempt);
+            SubscribeLocalEvent<DockingComponent, DoorCloseAttemptEvent>(OnDoorCloseAttempt);
+        }
+
+        // Won't allow users to override door controls
+        private void OnDoorOpenAttempt(EntityUid uid, DockingComponent component, DoorOpenAttemptEvent args)
+        {
+            args.Cancel();
+        }
```

Comment:
this also coveres crowbarring open unpowered airlocks. i dunno, should we maybe specialcase that?

Reply:
Could make prying call Open / Close directly and anything that wishes to block it can subscribe to BeforeDoorPryEvent?